sys/psa_crypto/psa_crypto.c

```diff
@@ -0,0 +1,1856 @@
+
+    *key = PSA_KEY_ID_NULL;
+    /* TODO: Finish persistent key storage */
+    /* TODO: Finish SE key storage with transaction */
+
+    if (status == PSA_SUCCESS) {
+        *key = slot->attr.id;
+        status = psa_unlock_key_slot(slot);
+    }
+    else {
+        (void)slot;
+    }
+
+    (void)driver;
+    return status;
+}
+
+/**
+ * @brief   Abort key creation and clean up in case of failure
+ *
+ * @param   slot    Slot that the key has been written to
+ * @param   driver  SE driver, in case the key creation took place on a secure element
+ */
+static void psa_fail_key_creation(psa_key_slot_t *slot, psa_se_drv_data_t *driver)
+{
+    (void)driver;
+    if (slot == NULL) {
+        return;
+    }
+    /* TODO: Destroy key in secure element (see mbedtls code) */
+    /* TODO: Secure Element stop transaction */
+    psa_wipe_key_slot(slot);
+}
+
+psa_status_t psa_copy_key(psa_key_id_t source_key,
+                          const psa_key_attributes_t *attributes,
+                          psa_key_id_t *target_key)
+{
+    (void)source_key;
+    (void)attributes;
+    (void)target_key;
+    return PSA_ERROR_NOT_SUPPORTED;
+}
+
+psa_status_t psa_destroy_key(psa_key_id_t key)
+{
+    psa_status_t status;
+    psa_key_slot_t *slot;
+
+    status = psa_get_and_lock_key_slot(key, &slot);
+    if (status != PSA_SUCCESS) {
+        return status;
+    }
+    if (slot->lock_count > 1) {
+        return PSA_ERROR_CORRUPTION_DETECTED;
+    }
+
+    return psa_wipe_key_slot(slot);
+}
+
+psa_status_t psa_export_key(psa_key_id_t key,
+                            uint8_t *data,
+                            size_t data_size,
+                            size_t *data_length)
+{
+    (void)key;
+    (void)data;
+    (void)data_size;
+    (void)data_length;
+    return PSA_ERROR_NOT_SUPPORTED;
+}
+
+/**
+ * @brief   Export asymmetric public key that is stored in local memory
+ *
+ *          See @ref psa_export_public_key
+ *
+ * @param   key_buffer
+ * @param   key_buffer_size
+ * @param   data
+ * @param   data_size
+ * @param   data_length
+ *
+ * @return  @ref PSA_SUCCESS
+ *          @ref PSA_ERROR_INVALID_ARGUMENT
+ */
+static psa_status_t psa_builtin_export_public_key( const uint8_t *key_buffer,
+                                                   size_t key_buffer_size,
+                                                   uint8_t *data,
+                                                   size_t data_size,
+                                                   size_t *data_length)
+{
+    if (key_buffer_size == 0 || data_size == 0) {
+        return PSA_ERROR_INVALID_ARGUMENT;
+    }
+    /** Some implementations and drivers can generate a public key from existing private key
+     * material. This implementation does not support the recalculation of a public key, yet.
+     * It requires the key to already exist in local memory and just copies it into the data
+     * output. */
+    memcpy(data, key_buffer, key_buffer_size);
```

Comment:
`data_size` is not checked! This might overflow the `data` buffer if `key_buffer_size` is greater than `data_size`.

Reply:
Done